src/NUnitFramework/framework/AssertionExtensions.cs

```diff
@@ -0,0 +1,39 @@
+#nullable enable
+
+namespace NUnit.Framework
+{
+    /// <summary>
+    /// Extension methods to help with Assert and nullability.
+    /// </summary>
+    public static class AssertionExtensions
+    {
+        /// <summary>
+        /// Assert that this source is not null.
+        /// </summary>
+        /// <typeparam name="T">The type of instance being handled.</typeparam>
+        /// <param name="source">The source instance to check for null.</param>
+        /// <returns>The source instance, or throws when null.</returns>
+        public static T AssertNotNull<T>(this T? source)
+            where T : class
+        {
+            Assert.NotNull(source);
+
+            return source;
+        }
```

Comment:
I just noticed these new extension method Assert APIs. This is something we should also discuss. If we add these, we should add `T : struct` copies of these methods.

Reply:
If we change the language version to C#9 we can drop the `T : class` constraint and it will work for both.

Reply:
The return type would not be as convenient if we did that though. `T?` means "defaultable" rather than "nullable" if there is no class or struct constraint.

Having to do `nullableValueType.AssertNotNull().Value` would be unpleasant to me since we can avoid making the `.Value` part (logically an assertion itself)  necessary.

Reply:
I stand corrected. Thanks. Disappointing that even though the return type is `T` it still returns `T?` I thought C#9 fixed this. 

As constraints are still not part of the signature we cannot have two versions, one with `where T : class` and one with `where T : struct`. However, I managed to make the 2nd by explicitly declaring it as:

        public static T AssertNotNull<T>(this Nullable<T> source)
            where T : struct

Reply:
We actually can have both versions because the signature is still different for the `T : struct` version when the parameter is `T?`.

The reason that C# 9 didn't fix this the way people hoped was that there is no sequence of bytes the compiler can emit that would work that way. The .NET runtime would have to gain a new feature and have a new way to encode ECMA-335 metadata that would not be compatible with previous runtimes and versions.

Reply:
You are correct again. Thanks.

Reply:
(The ability to *resolve* compiler overloads like this is a newish C# compiler feature IIRC, but IIRC it was backported to all previous language versions. Don't quote me on this.)

Reply:
No worries, glad I can help.